csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,502 @@
+        name = "ToString" and
+        input = TQualifierSummaryInput() and
+        inputAp = ContentStack::element() and
+        output = TReturnSummaryOutput() and
+        outputAp = ContentStack::empty() and
+        preservesValue = false
+        or
+        exists(int i, Type t |
+          name.regexpMatch("Append(Format|Line)?") and
+          t = m.getParameter(i).getType() and
+          input = TArgumentSummaryInput(i) and
+          inputAp = ContentStack::empty() and
+          output = [TQualifierSummaryOutput().(TSummaryOutput), TReturnSummaryOutput()] and
+          outputAp = ContentStack::element() and
+          preservesValue = true
+        |
+          t instanceof StringType or
+          t instanceof ObjectType
+        )
+      )
+    }
+
+    override predicate hasSummary(
+      SourceDeclarationCallable c, SummaryInput input, ContentStack inputStack,
+      SummaryOutput output, ContentStack outputStack, boolean preservesValue
+    ) {
+      (
+        this.constructorFlow(c, input, inputStack, output, outputStack) and
+        preservesValue = true
+        or
+        this.methodFlow(c, input, inputStack, output, outputStack, preservesValue)
+      )
+    }
+
+    override predicate clearsContent(
+      SourceDeclarationCallable c, SummaryInput input, Content content
+    ) {
+      c = this.getClass().getAMethod("Clear") and
+      input = TQualifierSummaryInput() and
+      content instanceof ElementContent
+    }
+  }
+
+  private import semmle.code.csharp.frameworks.System
+
+  /** Data flow for `System.Lazy`. */
+  class LazyDataFlow extends FrameworkDataFlow {
+    LazyDataFlow() { this = "System.Lazy" }
+
+    private SystemLazyClass getClass() { any() }
+
+    override predicate hasSummary(
+      SourceDeclarationCallable c, SummaryInput input, ContentStack inputStack,
+      SummaryOutput output, ContentStack outputStack, boolean preservesValue
+    ) {
+      preservesValue = true and
+      exists(SystemFuncDelegateType t, int i | t.getNumberOfTypeParameters() = 1 |
+        c.(Constructor).getDeclaringType() = this.getClass() and
+        c.getParameter(i).getType().getSourceDeclaration() = t and
+        input = SummaryInput::delegate(c, i) and
+        inputStack = ContentStack::empty() and
+        output = TReturnSummaryOutput() and
+        outputStack = ContentStack::property(this.getClass().getValueProperty())
+      )
+      or
+      preservesValue = false and
+      c = this.getClass().getValueProperty().getGetter() and
+      input = TQualifierSummaryInput() and
+      inputStack = ContentStack::empty() and
+      output = TReturnSummaryOutput() and
+      outputStack = ContentStack::empty()
+    }
+  }
+
+  private import semmle.code.csharp.frameworks.system.Data
+  private import semmle.code.csharp.frameworks.system.data.SqlClient
+
+  class SqlClientDataFlow extends FrameworkDataFlow {
+    SqlClientDataFlow() { this = "SqlClientDataFlow" }
```

Comment:
Assuming the string `"SqlClientDataFlow"` and `class SqlClientDataFlow` are never explicitly needed in the code, we could drop the configuration-based approach and write
```
// SinkNode is abstract and extends Node
class SqlClientSink extends SinkNode {
  override predicate hasSink(/* Node this, */ FlowSink sink) {
    ...
  }
}
```